### PR TITLE
Pad iteration for fileBased output

### LINKED ISSUE #265: Editing a data000X00 series
**Describe the bug**
Editing files in an *existing series* that is zero-padded file names fails due to a mismatch in file names (the zeroes are forgotten during write).

**To Reproduce**

Python:
```python
import openPMD

# $ ls samples/git-sample
# data00000100.h5  data00000200.h5  data00000300.h5  data00000400.h5  data00000500.h5

series = openPMD.Series("samples/git-sample/data%T.h5", openPMD.Access_Type.read_write)

series.iterations[42]
# <openPMD.Iteration of at t = '0.000000 s'>

series.flush()
```
```
HDF5-DIAG: Error detected in HDF5 (1.10.2) thread 0:
  #000: H5F.c line 511 in H5Fopen(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #001: H5Fint.c line 1450 in H5F_open(): unable to open file: time = Fri Jun 15 00:10:23 2018
, name = 'samples/git-sample/data100.h5', tent_flags = 1
    major: File accessibilty
    minor: Unable to open file
  #002: H5FD.c line 733 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #003: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = 'samples/git-sample/data100.h5', errno = 2, error message = 'No such file or directory', flags = 1, o_flags = 2
    major: File accessibilty
    minor: Unable to open file
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-8118d0dafbde> in <module>()
----> 1 series.flush()

RuntimeError: Failed to open HDF5 file samples/git-sample/data100.h5
```

**Expected behavior**
It should actually try to access to the `data00000100.h5` file but it skips the leading zeros.

**Software Environment:**
 - version of openPMD-api: 0.2.1-dev as of 402698012d2587a716b1273bb31333e86e38d4e5
 - installed openPMD-api via: from source with gcc 6.3.0
 - operating system: Debian GNU/Linux 9.4 (stretch)
 - machine: my thinkpad laptop
 - name and version of Python implementation:  CPython 3.6.5
 - version of HDF5: 1.10.2
 - version of ADIOS1: 1.13.1
 - name and version of MPI: OpenMPI 3.1.0

### PR DESCRIPTION
**Pull request by anokfireball**
Closes #265.

### CODE REVIEW

File: src/Series.cpp [hunk 1]
```diff
@@ -85,145 +97,105 @@ Series::Series(std::string filepath,
...
-    switch( at )
+        if( regexMatch[3].rm_so == -1 || regexMatch[3].rm_eo == -1 )
+            throw std::runtime_error("Can not determine iterationFormat (postfix) from filename " + input->name);
+        else
+            input->filenamePrefix = std::string(&input->name[regexMatch[3].rm_so], regexMatch[3].rm_eo);
+    }
+
+    regfree(&pattern);
+#else
+    std::regex pattern("(.*)%(0\\d+)?T(.*)");
```
Comment (anokfireball) [inline src/Series.cpp]: @ax3l this is a PoC of https://github.com/openPMD/openPMD-api/issues/265#issuecomment-397604092

File: src/Series.cpp [hunk 2]
```diff
@@ -85,145 +97,105 @@ Series::Series(std::string filepath,
...
-    if( std::string::npos != newName.find("%T") )
-        ie = IterationEncoding::fileBased;
+    input->format = determineFormat(input->name);
+
+#if defined(IS_GCC_48)
+    regex_t pattern;
+    if( regcomp(&pattern, "(.*)%(0\\d+)?T(.*)", REG_EXTENDED) )
+        throw std::runtime_error(std::string("Regex for iterationFormat '(.*)%(0\\d+)?T(.*)' can not be compiled!"));
+
+    uint8_t maxGroups = 4;
```
Comment (ax3l) [inline src/Series.cpp]: make this a `constexpr` to avoid declaring a variable length array

**Comment by ax3l** [comment]
@C0nsultant can you please add a test that demonstrates/checks the original issue (`read_write` to an existing, zero-padded series)?
It's a corner case that we might want to check permanently.

File: src/Series.cpp (hunk 2 above)
Comment (anokfireball) [reply src/Series.cpp]: I'll probably opt for an ```std::array< regmatch_t, 4 >```. This is C++, after all.

Comment (ax3l) [reply src/Series.cpp]: I still would keep the named constant as a separate line. Magic numbers are evil.
```C++
constexpr uint8_t maxGroups = 4u;
std::array< regmatch_t, maxGroups > regexMatch;
```

**Review by ax3l — APPROVED** [review]
Epic, very nice!

Don't forget to remove the `[WIP]` title and add your changes to the changelog when it is ready :)

File: include/openPMD/Series.hpp [hunk 3]
```diff
@@ -251,8 +251,8 @@ class Series : public Attributable
     std::shared_ptr< std::string > m_name;
     std::shared_ptr< Format > m_format;
 
-    std::string m_filenamePrefix;
-    std::string m_filenamePostfix;
-    int m_filenamePadding;
+    std::shared_ptr< std::string > m_filenamePrefix;
```
Comment (C0nsultant) [inline include/openPMD/Series.hpp]: Unit tests for these should be added.

File: include/openPMD/Series.hpp [hunk 4]
```diff
@@ -247,5 +250,9 @@ class Series : public Attributable
     std::shared_ptr< IterationEncoding > m_iterationEncoding;
     std::shared_ptr< std::string > m_name;
     std::shared_ptr< Format > m_format;
+
+    std::shared_ptr< std::string > m_filenamePrefix;
+    std::shared_ptr< std::string > m_filenamePostfix;
+    std::shared_ptr< int > m_filenamePadding;
```
Comment (anokfireball) [inline include/openPMD/Series.hpp:257]: Unit tests for these ```std::shared_ptrs``` should be added.

Comment (anokfireball) [reply include/openPMD/Series.hpp:257]: This comment can be dismissed, as these are implicitly tested in ```SerialIOTests```.
